telemetry/internal/appinsights/inmemorychannel.go

```diff
@@ -0,0 +1,311 @@
+type batchItem struct {
+	item    *contracts.Envelope
+	retries int
+}
+
+// inMemoryChannel stores events exclusively in memory.
+// Presently the only telemetry channel implementation available.
+type inMemoryChannel struct {
+	endpointAddr  string
+	batchSize     int
+	batchInterval time.Duration
+	errorLog      *log.Logger
+
+	collectChan chan *contracts.Envelope
+	flushChan   chan struct{}
+	retryChan   chan retryMessage
+
+	transmitter transmitter
+
+	// Use a context instead of a channel to
+	// allow propagating the cancellation to the
+	// transmitter and to the underlying HTTP client.
+	cancelCtx       context.Context
+	cancelCauseFunc context.CancelCauseFunc
+
+	throttled   atomic.Bool
+	closed      atomic.Bool
+	sendQueue   []*[]batchItem
+	sendQueueMu sync.Mutex
+	sendMu      sync.Mutex
+}
+
+type retryMessage struct {
+	throttled  bool
+	retryAfter time.Time
+	items      []batchItem
+}
+
+// newInMemoryChannel creates an inMemoryChannel instance and starts a background submission goroutine.
+func newInMemoryChannel(endpointUrl string, batchSize int, batchInterval time.Duration, httpClient *http.Client, errorLog *log.Logger) *inMemoryChannel {
+	// Set up the channel
+	channel := &inMemoryChannel{
+		endpointAddr:  endpointUrl,
+		batchSize:     batchSize,
+		batchInterval: batchInterval,
+		errorLog:      errorLog,
+		collectChan:   make(chan *contracts.Envelope),
+		flushChan:     make(chan struct{}),
+		retryChan:     make(chan retryMessage),
+		transmitter:   newTransmitter(endpointUrl, httpClient),
+	}
+	channel.cancelCtx, channel.cancelCauseFunc = context.WithCancelCause(context.Background())
+	return channel
+}
+
+func (channel *inMemoryChannel) logf(format string, args ...any) {
+	if channel.errorLog != nil {
+		channel.errorLog.Printf(format, args...)
+	} else {
+		log.Printf(format, args...)
+	}
+}
+
+// Queues a single telemetry item
+func (channel *inMemoryChannel) send(item *contracts.Envelope) {
+	if item != nil && !channel.closed.Load() {
+		channel.collectChan <- item
+	}
+}
+
+// Forces the current queue to be sent
+func (channel *inMemoryChannel) flush() {
+	if channel.closed.Load() {
+		return
+	}
+	channel.flushChan <- struct{}{}
+}
+
+func (channel *inMemoryChannel) retry(throttled bool, retryAfter time.Time, items []batchItem) {
+	// Retry even if the channel is closed to allow for
+	// retrying items that were already sent.
+	channel.retryChan <- retryMessage{throttled, retryAfter, items}
+}
+
+var errStopped = errors.New("client stopped")
+var errClosed = errors.New("client closed")
+
+func (channel *inMemoryChannel) stop() {
+	if channel.closed.Load() {
+		return
+	}
+
+	channel.closed.Store(true)
+	channel.cancelCauseFunc(errStopped)
+}
+
+// Flushes and tears down the submission goroutine and closes internal
+// channels.  Returns a channel that is closed when all pending telemetry
+// items have been submitted and it is safe to shut down without losing
+// telemetry.
```

Comment:
Doesn't seem accurate?

```suggestion
// close flushes and tears down the submission goroutine and closes internal
// channels. Returns when all pending telemetry items have been submitted
// (it is then safe to shut down without losing telemetry) or when
// the context is canceled.
```